Import necessary libraries

In [1]:
import pandas as pd
import numpy as np

**1.** Fetch dataset. In order to have possibility to fetch it by URL, I have uploaded dataset to my GitHub.

Dataset is about aid from different countries for Ukraine. Original dataset is taken from https://www.kaggle.com/datasets/jainaru/financial-aid-to-ukraine.

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/YaroslavDrovolskyi/applied-application-of-neural-networks/main/lab1/Financial%20Allocations.csv")

**2.** Print dataset

In [3]:
df

Country  EU member  Financial allocations($ billion)  \
0        Australia          0                          0.728552   
1          Austria          1                          0.191309   
2          Belgium          1                          0.000000   
3         Bulgaria          1                          3.695839   
4           Canada          0                          0.000000   
5          Croatia          1                          0.001071   
6           Cyprus          1                          0.000000   
7   Czech Republic          1                          0.131348   
8          Denmark          1                          0.000000   
9          Estonia          1                         29.656888   
10         Finland          1                          0.856477   
11          France          1                          1.507460   
12         Germany          1                          0.000000   
13          Greece          1                          0.000000   
14         Hungary          1                          0.010197   
15         Iceland          0                          0.042856   
16         Ireland          1                          0.439274   
17           Italy          1                          5.180550   
18           Japan          0                          0.043927   
19          Latvia          1                          0.044463   
20       Lithuania          1                          0.001393   
21      Luxembourg          1                          0.000000   
22           Malta          1                          0.766694   
23     Netherlands          1                          0.001824   
24     New Zealand          0                          0.826775   
25          Norway          0                          1.013623   
26          Poland          1                          0.000000   
27        Portugal          1                          0.310000   
28     South Korea          0                          0.000000   
29         Romania          1                          0.005357   
30        Slovakia          1                          0.000000   
31        Slovenia          1                          0.556807   
32           Spain          1                          0.367611   
33          Sweden          1                          0.110382   
34     Switzerland          0                          0.000000   
35          Turkey          0                          3.510938   
36  United Kingdom          0                         23.000000   
37   United States          0                          0.000000   
38           China          0                          0.000000   
39          Taiwan          0                          0.000000   
40           India          0                               NaN   

    Humanitarian allocations($ billion)  Military allocations($ billion)  \
0                              0.069499                         0.003587   
1                              0.053399                         0.923287   
2                              0.001041                         0.262288   
3                              0.397457                              NaN   
4                              0.086783                         0.203566   
5                              0.003214                         0.000000   
6                              0.057240                         1.378977   
7                              0.261334                         5.116618   
8                              0.052539                         0.524732   
9                              2.449756                         0.000000   
10                             0.387744                         2.883296   
11                             3.263849                        10.752226   
12                             0.000000                         0.143237   
13                             0.057910                         0.000000   
14                             0.011962                 

**3.** Check if any value in dataframe is `NaN`.

In [4]:
df.isnull().any().any()

True

**4.** Dataset has `NaN` values, so need to replace them by `10`. Print dataset with replaces `NaN`s

In [5]:
df.fillna(10)

Country  EU member  Financial allocations($ billion)  \
0        Australia          0                          0.728552   
1          Austria          1                          0.191309   
2          Belgium          1                          0.000000   
3         Bulgaria          1                          3.695839   
4           Canada          0                          0.000000   
5          Croatia          1                          0.001071   
6           Cyprus          1                          0.000000   
7   Czech Republic          1                          0.131348   
8          Denmark          1                          0.000000   
9          Estonia          1                         29.656888   
10         Finland          1                          0.856477   
11          France          1                          1.507460   
12         Germany          1                          0.000000   
13          Greece          1                          0.000000   
14         Hungary          1                          0.010197   
15         Iceland          0                          0.042856   
16         Ireland          1                          0.439274   
17           Italy          1                          5.180550   
18           Japan          0                          0.043927   
19          Latvia          1                          0.044463   
20       Lithuania          1                          0.001393   
21      Luxembourg          1                          0.000000   
22           Malta          1                          0.766694   
23     Netherlands          1                          0.001824   
24     New Zealand          0                          0.826775   
25          Norway          0                          1.013623   
26          Poland          1                          0.000000   
27        Portugal          1                          0.310000   
28     South Korea          0                          0.000000   
29         Romania          1                          0.005357   
30        Slovakia          1                          0.000000   
31        Slovenia          1                          0.556807   
32           Spain          1                          0.367611   
33          Sweden          1                          0.110382   
34     Switzerland          0                          0.000000   
35          Turkey          0                          3.510938   
36  United Kingdom          0                         23.000000   
37   United States          0                          0.000000   
38           China          0                          0.000000   
39          Taiwan          0                          0.000000   
40           India          0                         10.000000   

    Humanitarian allocations($ billion)  Military allocations($ billion)  \
0                              0.069499                         0.003587   
1                              0.053399                         0.923287   
2                              0.001041                         0.262288   
3                              0.397457                        10.000000   
4                              0.086783                         0.203566   
5                              0.003214                         0.000000   
6                              0.057240                         1.378977   
7                              0.261334                         5.116618   
8                              0.052539                         0.524732   
9                              2.449756                         0.000000   
10                             0.387744                         2.883296   
11                             3.263849                        10.752226   
12                             0.000000                         0.143237   
13                             0.057910                         0.000000   
14                             0.011962                 

**5.** Delete rows with `NaN` values. (Drop the rows where at least one element is `NaN`).

In [6]:
df=df.dropna()
df

Country  EU member  Financial allocations($ billion)  \
0        Australia          0                          0.728552   
1          Austria          1                          0.191309   
2          Belgium          1                          0.000000   
4           Canada          0                          0.000000   
5          Croatia          1                          0.001071   
6           Cyprus          1                          0.000000   
7   Czech Republic          1                          0.131348   
8          Denmark          1                          0.000000   
9          Estonia          1                         29.656888   
10         Finland          1                          0.856477   
11          France          1                          1.507460   
12         Germany          1                          0.000000   
13          Greece          1                          0.000000   
14         Hungary          1                          0.010197   
15         Iceland          0                          0.042856   
16         Ireland          1                          0.439274   
17           Italy          1                          5.180550   
18           Japan          0                          0.043927   
20       Lithuania          1                          0.001393   
21      Luxembourg          1                          0.000000   
22           Malta          1                          0.766694   
23     Netherlands          1                          0.001824   
24     New Zealand          0                          0.826775   
25          Norway          0                          1.013623   
26          Poland          1                          0.000000   
28     South Korea          0                          0.000000   
29         Romania          1                          0.005357   
30        Slovakia          1                          0.000000   
31        Slovenia          1                          0.556807   
32           Spain          1                          0.367611   
33          Sweden          1                          0.110382   
34     Switzerland          0                          0.000000   
35          Turkey          0                          3.510938   
36  United Kingdom          0                         23.000000   
37   United States          0                          0.000000   

    Humanitarian allocations($ billion)  Military allocations($ billion)  \
0                              0.069499                         0.003587   
1                              0.053399                         0.923287   
2                              0.001041                         0.262288   
4                              0.086783                         0.203566   
5                              0.003214                         0.000000   
6                              0.057240                         1.378977   
7                              0.261334                         5.116618   
8                              0.052539                         0.524732   
9                              2.449756                         0.000000   
10                             0.387744                         2.883296   
11                             3.263849                        10.752226   
12                             0.000000                         0.143237   
13                             0.057910                         0.000000   
14                             0.011962                         0.017934   
15                             0.060264                         0.004125   
16                             0.345916                         1.071400   
17                             1.094119                         0.069452   
18                             0.008164                         0.403534   
20                             0.019430                         0.171816   
21                             0.002218                         0.000000   
22

**6.** Create new column ('Total allocations')  as a sum of three existing columns

In [8]:
df.loc[:, 'Total allocations'] = df['Financial allocations($ billion)'] + df['Humanitarian allocations($ billion)'] + df['Military allocations($ billion)']

In [9]:
df

Country  EU member  Financial allocations($ billion)  \
0        Australia          0                          0.728552   
1          Austria          1                          0.191309   
2          Belgium          1                          0.000000   
4           Canada          0                          0.000000   
5          Croatia          1                          0.001071   
6           Cyprus          1                          0.000000   
7   Czech Republic          1                          0.131348   
8          Denmark          1                          0.000000   
9          Estonia          1                         29.656888   
10         Finland          1                          0.856477   
11          France          1                          1.507460   
12         Germany          1                          0.000000   
13          Greece          1                          0.000000   
14         Hungary          1                          0.010197   
15         Iceland          0                          0.042856   
16         Ireland          1                          0.439274   
17           Italy          1                          5.180550   
18           Japan          0                          0.043927   
20       Lithuania          1                          0.001393   
21      Luxembourg          1                          0.000000   
22           Malta          1                          0.766694   
23     Netherlands          1                          0.001824   
24     New Zealand          0                          0.826775   
25          Norway          0                          1.013623   
26          Poland          1                          0.000000   
28     South Korea          0                          0.000000   
29         Romania          1                          0.005357   
30        Slovakia          1                          0.000000   
31        Slovenia          1                          0.556807   
32           Spain          1                          0.367611   
33          Sweden          1                          0.110382   
34     Switzerland          0                          0.000000   
35          Turkey          0                          3.510938   
36  United Kingdom          0                         23.000000   
37   United States          0                          0.000000   

    Humanitarian allocations($ billion)  Military allocations($ billion)  \
0                              0.069499                         0.003587   
1                              0.053399                         0.923287   
2                              0.001041                         0.262288   
4                              0.086783                         0.203566   
5                              0.003214                         0.000000   
6                              0.057240                         1.378977   
7                              0.261334                         5.116618   
8                              0.052539                         0.524732   
9                              2.449756                         0.000000   
10                             0.387744                         2.883296   
11                             3.263849                        10.752226   
12                             0.000000                         0.143237   
13                             0.057910                         0.000000   
14                             0.011962                         0.017934   
15                             0.060264                         0.004125   
16                             0.345916                         1.071400   
17                             1.094119                         0.069452   
18                             0.008164                         0.403534   
20                             0.019430                         0.171816   
21                             0.002218                         0.000000   
22

**7.** Make generalized description of dataset

In [10]:
df.describe()

EU member  Financial allocations($ billion)  \
count  35.000000                         35.000000   
mean    0.685714                          1.970037   
std     0.471008                          6.227082   
min     0.000000                          0.000000   
25%     0.000000                          0.000000   
50%     1.000000                          0.042856   
75%     1.000000                          0.747623   
max     1.000000                         29.656888   

       Humanitarian allocations($ billion)  Military allocations($ billion)  \
count                            35.000000                        35.000000   
mean                              0.404842                         2.520189   
std                               0.793996                         7.917894   
min                               0.000000                         0.000000   
25%                               0.011897                         0.011189   
50%                               0.060264                         0.203566   
75%                               0.381136                         1.225188   
max                               3.263849                        46.152000   

       Share in EU allocations($ billion)  Total allocations  
count                           35.000000          35.000000  
mean                             0.587774           4.895068  
std                              1.208496          13.126261  
min                              0.000000           0.002218  
25%                              0.000000           0.096013  
50%                              0.107600           0.737384  
75%                              0.524792           3.803397  
max                              5.689608          71.858826

As we see, each country, **on average**, has allocated 4.895068 bn USD for Ukraine as support.

Minimum donation is 0.002218 bn USD, maximal is 71.858826 bn USD.


**8.** Make analysis for EU and non-EU countries






In [20]:
df.groupby(['EU member']).agg({'Financial allocations($ billion)':[min, max, sum, 'mean'],
                               'Humanitarian allocations($ billion)':[min, max, sum, 'mean'],
                               'Military allocations($ billion)':[min, max, sum, 'mean'],
                               'Total allocations':[min, max, sum, 'mean']})

<ipython-input-20-f7d240ff313c>:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  df.groupby(['EU member']).agg({'Financial allocations($ billion)':[min, max, sum, 'mean'],
<ipython-input-20-f7d240ff313c>:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df.groupby(['EU member']).agg({'Financial allocations($ billion)':[min, max, sum, 'mean'],
<ipython-input-20-f7d240ff313c>:1: FutureWarning: The provided callable <built-in function sum> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  df.groupby(['E

Financial allocations($ billion)                                  \
                                       min        max        sum      mean   
EU member                                                                    
0                                      0.0  23.000000  29.166671  2.651516   
1                                      0.0  29.656888  39.784641  1.657693   

          Humanitarian allocations($ billion)                                \
                                          min       max       sum      mean   
EU member                                                                     
0                                    0.002284  2.706826  4.438970  0.403543   
1                                    0.000000  3.263849  9.730485  0.405437   

          Military allocations($ billion)                                  \
                                      min        max        sum      mean   
EU member                                                                   
0                                     0.0  46.152000  56.594677  5.144971   
1                                     0.0  10.752226  31.611938  1.317164   

          Total allocations                                  
                        min        max        sum      mean  
EU member                                                    
0                  0.002284  71.858826  90.200317  8.200029  
1                  0.002218  32.106644  81.127064  3.380294

As we see, average non-EU country has donated 8.2 bn USD for Ukraine, and average EU country - 3.38 bn USD.

Minimum not-EU country allocation is 0.002284 bn USD, maximal is 71.858826 bn USD.
Minimum EU country allocation is 0.002218 bn USD, maximal is 32.106644 bn USD.

So, EU countries have allocated 81.12 bn USD, but not-EU countries - 90.2 bn USD.